api/src/main/java/org/openmrs/module/aijar/dataintegrity/InvalidTBEncounters.java

```diff
@@ -0,0 +1,200 @@
+		
+		//For each patient enrolled in the TB program, check if the current identifier is missing
+		for (Patient patient : patientList) {
+			List<Obs> patientObs = Context.getObsService().getObservationsByPersonAndConcept(patient.getPerson(), Dictionary.getConcept(tbIdentifierConceptUuid));
+			boolean identifierFound = false;
+			if (patientObs.size() == 0) {
+				identifierFound = false;
+			} else {
+				for (Obs obs : patientObs) {
+					if (obs.getValueText() != null ) {
+						PatientProgram pp = Context.getProgramWorkflowService().getPatientPrograms(patient, tbProgram, null, null, null, null, false).get(0);
+						Date treatmentStartDate = obs.getEncounter().getEncounterDatetime();
+						
+						//Compare the TB treatment start date matching the given Obs with the tb program enrollment date. 
+						//If the TB treatment date falls on or after the program enrollment date, consider this a valid identifier 
+						if (treatmentStartDate.equals(pp.getDateEnrolled()) || treatmentStartDate.after(pp.getDateEnrolled())){
+							identifierFound = true;
+						}
+					}
+				}				
+			}
+			
+			if (!identifierFound) {
+				RuleResult<Patient> ruleResult = new RuleResult<>();
+				ruleResult.setActionUrl("coreapps/patientdashboard/patientDashboard.page?patientId=" + patient.getUuid());
+				ruleResult.setNotes("Client# " + patient.getId() + ", has missing " + identifierTitle + " identifier");
+				ruleResult.setEntity(patient);
+				
+				ruleResults.add(ruleResult);				
+			}
+		}
+		
+		return ruleResults;		
+	}	
+		
+	/**
+	 * Patients with no final outcome after 9 months at start of treatment 
+	 * @return List<RuleResult<Patient>>
+	 */
+	public List<RuleResult<Patient>> patientsWithNoFinalOutcomeNineMonthsAfterStartOfTreatment(DateTime date) {
+		log.info("Executing rule to find patients with no final TB Outcome 9 months after start of treatment");
+		
+		String queryString = "SELECT pp.patient FROM PatientProgram pp WHERE pp.voided = 0 AND pp.dateEnrolled <= :date AND pp.outcome IS NULL AND pp.program.uuid = '" + Programs.TB_PROGRAM.uuid() + "'";
+		Query query = getSession().createQuery(queryString);
+        DateTime nineMonthsBefore = date.plusMonths(-9);
+        query.setParameter("date", nineMonthsBefore.toDate());
+	
+		List<Patient> patientsInProgramList = query.list();
+				
+		Set<Patient> uniquePatientList = new HashSet<Patient>(patientsInProgramList);
+		log.info("There are " + uniquePatientList.size() + " Patients with no final TB Outcome 9 months after start of treatment");
+		
+		List<RuleResult<Patient>> ruleResults = new ArrayList<>();
+		for (Patient patient : uniquePatientList) {
+			RuleResult<Patient> ruleResult = new RuleResult<>();
+			ruleResult.setActionUrl("coreapps/patientdashboard/patientDashboard.page?patientId=" + patient.getUuid());
+			ruleResult.setNotes("Patient #" + patient.getId() + " has no final TB Outcome 9 months after start of treatment");
```

Comment:
@moshonk one of the TB identifiers (if not available) then fall back to the OpenMRS ID for the patient as this is searchable not the numeric database ID